Tribes/src/app/_components/building-type/building-type.component.ts

```diff
@@ -91,13 +101,88 @@ export class BuildingTypeComponent implements OnInit {
 
   addNewBuilding() {
     if (this.goldAmount > COST_NEW_BUILDING) {
+      this.goldAmount = this.goldAmount - COST_NEW_BUILDING;
       this.buildingService.addNewBuilding(this.type);
     } else {
       this.errorMessage =  'You do not have enough money.';
     }
   }
 
   checkIfBuildingIsUpgradeable(level: number): boolean {
-    return this.getNumberOfBuildingsByLevel(level) !== 0 && this.townhallLevel > level && this.goldAmount >= this.getUpgradingCost(level);
+    return this.getNumberOfBuildingsByLevel(level) !== 0
+            && this.townhallLevel > level
+            && this.goldAmount >= this.getUpgradingCost(level)
+            && !this.checkIfAllBuildingByLevelIsUnderConstruction(level);
+  }
+
+  checkUpgradingConditions(level: number): string {
```

Comment:
As a naming convention, I would rather use something like 
`isUpgradePossible`, which returns a boolean, or a `validateIfUpgradeBuildingIsPossible`. I am also not satisfied fully with these. Also, I would consider putting them into a service maybe? i.e. imagine if the requirement would be to "yeah, let's put the upgrade button out in another screen" - this logic can be isolated quite well.

Reply:
I have 2 functions: 
1. checkIfTownhallIsUpgrading() -> renamed to isUpgradePossible(): this returns a boolean
2. checkUpgradingConditions() returns a string, that I use by creating the error message, I renamed it to validateUpgradingConditions() Is it better?

I was thinking to put all this functions into the building.service, but to do this, I need to put a lot of other functions to the service as well and I should use a lot of parameters (like townhallLevel, type, goldAmount), because the service doesn't know these. Should I put it into the service even so?